### PR TITLE
[generate] Drop attention mask early without padding

### PR DESCRIPTION
**Pull request by Cyrilvallez**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48814) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48814)

# What does this PR do?

As per the title. Supersedes https://github.com/huggingface/transformers/pull/47951 and improves https://github.com/huggingface/transformers/pull/48803 with a more general implem, i.e. we always do the same, not something different based on generation method, compilation etc

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -727,6 +726,11 @@ def prepare_inputs_for_generation(
                 mm_token_type_ids=model_inputs.get("mm_token_type_ids"),
                 is_first_iteration=is_first_iteration,
             )
+            if attention_mask is not None:
+                if isinstance(attention_mask, dict):
+                    attention_mask = {k: v.contiguous() for k, v in attention_mask.items()}
+                else:
+                    attention_mask = attention_mask.contiguous()
```
Comment (Cyrilvallez) [inline src/transformers/generation/utils.py]: To avoid recompilation due to stride (we also do it on other inputs). Almost all the time it's already contiguous and this is a no-op, but for some reason the compile test was failing on Chameleon and GPTBigCode due to stride

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -727,6 +726,10 @@ def prepare_inputs_for_generation(
                 mm_token_type_ids=model_inputs.get("mm_token_type_ids"),
                 is_first_iteration=is_first_iteration,
             )
+            if isinstance(attention_mask, dict):
+                attention_mask = {k: v.contiguous() if v is not None else None for k, v in attention_mask.items()}
+            else:
+                attention_mask = attention_mask.contiguous() if attention_mask is not None else None
```
Comment (Cyrilvallez) [inline src/transformers/generation/utils.py:732]: To avoid recompilation due to stride (we also do it on other inputs). Almost all the time it's already contiguous and this is a no-op, but for some reason the compile test was failing on Chameleon and GPTBigCode due to stride

**Review by SunMarc — APPROVED** [review]
Wow nice, thanks for fixing the edge cases 🤩

**Comment by Cyrilvallez** [comment]
Failing test is remote generate code on the hub, I already merged https://huggingface.co/transformers-community/contrastive-search/discussions/5 to fix it
